client/unittest/ReportIssueCreator.test.ts

```diff
@@ -0,0 +1,291 @@
+        const builtinExtension: vscode.Extension<any> = {
+            id: 'something.builtin',
+            packageJSON: {
+                name: 'BuiltInThing',
+                isBuiltin: true,
+            },
+        };
+        api.setExtensions(razorClientExtension, builtinExtension, omniSharpExtension);
+        const issueCreator = getReportIssueCreator(api);
+
+        // Act
+        const extensions = issueCreator.getInstalledExtensions();
+
+        // Assert
+        assert.deepEqual(extensions, [omniSharpExtension, razorClientExtension]);
+    });
+
+    it('sortExtensions OmniSharp is before Razor', async () => {
+        // Arrange
+        const api = createTestVSCodeApi();
+        const issueCreator = getReportIssueCreator(api);
+
+        // Act
+        const value = issueCreator.sortExtensions(omniSharpExtension, razorClientExtension);
+
+        // Assert
+        assert.equal(-1, value);
+    });
+
+    it('sortExtensions Razor is after OmniSharp', async () => {
+        // Arrange
+        const api = createTestVSCodeApi();
+        const issueCreator = getReportIssueCreator(api);
+
+        // Act
+        const value = issueCreator.sortExtensions(razorClientExtension, omniSharpExtension);
+
+        // Assert
+        assert.equal(1, value);
+    });
+
+    it('sortExtensions OmniSharp is identical to OmniSharp', async () => {
+        // Arrange
+        const api = createTestVSCodeApi();
+        const issueCreator = getReportIssueCreator(api);
+
+        // Act
+        const value = issueCreator.sortExtensions(omniSharpExtension, omniSharpExtension);
+
+        // Assert
+        assert.equal(0, value);
+    });
+
+    it('generateExtensionTable returns all non-builtin extensions in string format', async () => {
+        // Arrange
+        const api = createTestVSCodeApi();
+        const builtinExtension: vscode.Extension<any> = {
+            id: 'something.builtin',
+            packageJSON: {
+                name: 'BuiltInThing',
+                version: 'ShouldNotShowUp',
+                isBuiltin: true,
+            },
+        };
+        api.setExtensions(razorClientExtension, builtinExtension, omniSharpExtension);
+        const issueCreator = getReportIssueCreator(api);
+
+        // Act
+        const table = issueCreator.generateExtensionTable();
+
+        // Assert
+        assert.ok(table.indexOf(omniSharpExtension.packageJSON.version) > 0);
+        assert.ok(table.indexOf(razorClientExtension.packageJSON.version) > 0);
+        assert.ok(table.indexOf(builtinExtension.packageJSON.version) === -1);
+    });
+
+    it('generateExtensionTable can operate when 0 extensions', async () => {
+        // Arrange
+        const api = createTestVSCodeApi();
+        const issueCreator = getReportIssueCreator(api);
+
+        // Act & Assert
+        assert.doesNotThrow(() => issueCreator.generateExtensionTable());
+    });
+});
+
+class TestReportIssueCreator extends ReportIssueCreator {
```

Comment:
I find this pattern a bit odd. We're unit testing methods that are protected only so they can be unit tested. I know we have a habit of doing this in shipping packages where the API surface needs to be locked down, but it's possibly different here since nobody except for us sees these APIs at all.

Other ways we could structure this:

* Only unit test the naturally public API surface. If a behavior isn't observable via public APIs, then there's no reason to specify that behavior.
* Or, maybe make the methods `public`. Nothing apart from our code can even see it either way. Then we don't need the `TestReportIssueCreator` subclass.

I'm not suggesting this is a big problem! If you prefer to leave it as-is, I don't object. I'm just wondering if this is a case where some patterns are leaking out from C#/framework coding practices into a scenario where they aren't really applicable.

Reply:
> * Only unit test the naturally public API surface. If a behavior isn't observable via public APIs, then there's no reason to specify that behavior.

I try to avoid this approach because you end up writing super complex integration level tests that have to fill out every part of the API that is used. In particular, the VSCode case is extra concerning because a lot of the APIs can't be mocked or faked properly.



> * Or, maybe make the methods `public`. Nothing apart from our code can even see it either way. Then we don't need the `TestReportIssueCreator` subclass.

I made them protected instead of public to avoid production code misuses. Aka, if a new dev is typing dot and then getting a completion list of available methods I don't want them to be using methods that are only public for the sake of testing.